packages/voting/src/deposit.rs

```diff
@@ -19,107 +44,349 @@ pub enum DepositToken {
                 )?;
-                token_addr
+                // We don't assume here that the voting module has
+                // returned a valid token. Conversion of the unchecked
+                // denom into a checked one will do a `TokenInfo {}`
+                // query.
+                UncheckedDenom::Cw20(token_addr.into_string()).into_checked(deps)
             }
-        };
-        // Make an info query as a smoke test that we are indeed
-        // working with a token here. We can't turbofish this
-        // type. See <https://github.com/rust-lang/rust/issues/83701>
-        let _info: cw20::TokenInfoResponse = deps
-            .querier
-            .query_wasm_smart(token.clone(), &cw20::Cw20QueryMsg::TokenInfo {})?;
+        }?;
+
         Ok(CheckedDepositInfo {
-            token,
-            deposit,
-            refund_failed_proposals,
+            denom,
+            amount,
+            refund_policy,
         })
     }
 }
 
-pub fn get_deposit_msg(
-    info: &Option<CheckedDepositInfo>,
-    contract: &Addr,
-    sender: &Addr,
-) -> StdResult<Vec<CosmosMsg>> {
-    match info {
-        Some(info) => {
-            if info.deposit.is_zero() {
-                Ok(vec![])
+impl CheckedDepositInfo {
+    pub fn check_native_deposit_paid(&self, info: &MessageInfo) -> Result<(), DepositError> {
+        if let Self {
+            amount,
+            denom: CheckedDenom::Native(denom),
+            ..
+        } = self
+        {
+            // must_pay > may_pay. The method this is getting called
+            // in is accepting a deposit. It seems likely to me that
+            // if other payments are here it's a bug in a frontend and
+            // not an intentional thing.
+            let paid = must_pay(info, denom)?;
+            if paid != *amount {
+                Err(DepositError::InvalidDeposit {
+                    actual: paid,
+                    expected: *amount,
+                })
+            } else {
+                Ok(())
+            }
+        } else {
+            // Nothing to do if we're a cw20.
+            Ok(())
+        }
+    }
+
+    pub fn get_take_deposit_messages(
+        &self,
+        depositor: &Addr,
+        contract: &Addr,
+    ) -> StdResult<Vec<CosmosMsg>> {
+        let take_deposit_msg: Vec<CosmosMsg> = if let Self {
+            amount,
+            denom: CheckedDenom::Cw20(address),
+            ..
+        } = self
+        {
+            // into_checked() makes sure this isn't the case, but just for
+            // posterity.
+            if amount.is_zero() {
+                vec![]
             } else {
-                let transfer_msg = WasmMsg::Execute {
-                    contract_addr: info.token.to_string(),
+                vec![WasmMsg::Execute {
+                    contract_addr: address.to_string(),
                     funds: vec![],
                     msg: to_binary(&cw20::Cw20ExecuteMsg::TransferFrom {
```

Comment:
does the depositor create an allowance for this? where does that happen / is it triggered by the frontend?

Reply:
yes, triggered by frontend